### PR TITLE
Add serverside objects move events

### PR DESCRIPTION
**Pull request by FileEX**
We have ``onClientObjectMoveStart`` and ``onClientObjectMoveStop`` events - This PR adds these events also for the server side.

The ``stoppedByScript`` parameter determines whether the object has been stopped using stopObject or has reached its destination

### CODE REVIEW

File: Server/mods/deathmatch/logic/CObject.cpp [hunk 1]
```diff
@@ -346,9 +346,12 @@ void CObject::Move(const CPositionRotationAnimation& a_rMoveAnimation)
         SetPosition(positionRotation.m_vecPosition);
         SetRotation(positionRotation.m_vecRotation);
     }
+
+    CLuaArguments Arguments;
+    this->CallEvent("onObjectMoveStart", Arguments);
```
Comment (tederis) [inline Server/mods/deathmatch/logic/CObject.cpp]: Is `this->` really needed?

**Comment by FileEX** [comment]
I had to change the way the ``onObjectStop`` event was triggered. This must be synchronized via a packet on the client side because on the server side StopMovement is not called when the object reaches its destination as this happens on the client side.

I also added the ``stoppedByScript`` parameter for the ``onClientObjectMoveStop`` event on the client side.

File: Client/mods/deathmatch/logic/CClientGame.cpp [hunk 2]
```diff
@@ -2772,7 +2772,7 @@ void CClientGame::AddBuiltInEvents()
     m_Events.AddEvent("onClientObjectDamage", "loss, attacker", NULL, false);
     m_Events.AddEvent("onClientObjectBreak", "attacker", NULL, false);
     m_Events.AddEvent("onClientObjectMoveStart", "", NULL, false);
-    m_Events.AddEvent("onClientObjectMoveStop", "", NULL, false);
+    m_Events.AddEvent("onClientObjectMoveStop", "stoppedByScript", NULL, false);
```
Comment (TracerDS) [inline Client/mods/deathmatch/logic/CClientGame.cpp]: `nullptr` instead of `NULL`

File: Client/mods/deathmatch/logic/CDeathmatchObject.cpp [hunk 3]
```diff
@@ -89,7 +89,12 @@ void CDeathmatchObject::_StopMovement(bool a_bUnregister)
         delete m_pMoveAnimation;
         m_pMoveAnimation = NULL;
         CLuaArguments Arguments;
-        this->CallEvent("onClientObjectMoveStop", Arguments, true);
+        Arguments.PushBoolean(bStoppedByScript);
+
+        CallEvent("onClientObjectMoveStop", Arguments, true);
+
+        if (!IsLocalEntity())
+            SendObjectStopPacket(bStoppedByScript);
```
Comment (TracerDS) [inline Client/mods/deathmatch/logic/CDeathmatchObject.cpp]: Early return

File: Client/mods/deathmatch/logic/CDeathmatchObject.cpp [hunk 4]
```diff
@@ -89,7 +89,12 @@ void CDeathmatchObject::_StopMovement(bool a_bUnregister)
         delete m_pMoveAnimation;
         m_pMoveAnimation = NULL;
```
Comment (TracerDS) [inline Client/mods/deathmatch/logic/CDeathmatchObject.cpp]: `nullptr` instead of `NULL`

File: Server/mods/deathmatch/logic/CGame.cpp [hunk 5]
```diff
@@ -1546,6 +1552,8 @@ void CGame::AddBuiltInEvents()
     m_Events.AddEvent("onPlayerVoiceStop", "", NULL, false);
 
     // Object events
+    m_Events.AddEvent("onObjectMoveStart", "", NULL, false);
+    m_Events.AddEvent("onObjectMoveStop", "stoppedByScript", NULL, false);
```
Comment (TracerDS) [inline Server/mods/deathmatch/logic/CGame.cpp]: `nullptr` instead of `NULL`

**Review by TracerDS — CHANGES_REQUESTED** [review]
Good job

File: Client/mods/deathmatch/logic/CDeathmatchObject.cpp (hunk 4 above)
Comment (FileEX) [reply Client/mods/deathmatch/logic/CDeathmatchObject.cpp]: This is not my code. There are a lot of NULLs throughout the file, as well as throughout the project. Better to do a separate PR with refactoring

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CDeathmatchObject.cpp]: > This is not my code. There are a lot of NULLs throughout the file, as well as throughout the project. Better to do a separate PR with refactoring

Well, you are changing the file already so the change would not be significant

Comment (FileEX) [reply Client/mods/deathmatch/logic/CDeathmatchObject.cpp]: Okay, so let's do it

**Comment by TheNormalnij** [comment]
A packet from a client to notify server about this event is overkill.
onObjectMoveStart and onObjectMoveStop are pointless on server, because server has full control over created objects in scripts, and setTimer replaces onObjectMoveStop.
This realization still request setTimer, because onObjectMoveStop doesn't work on empty server. This can cause frustrating behavior and strange bugs in server scripts

**Comment by FileEX** [comment]
So how to improve it?

**Pull request closed without merging**